internal/dag/gatewayapi_processor.go

```diff
@@ -1502,6 +1786,89 @@ func (p *GatewayAPIProcessor) clusterRoutes(routeNamespace string, matchConditio
+	totalWeight := uint32(0)
+
+	if len(backendRefs) == 0 {
+		routeAccessor.AddCondition(gatewayapi_v1beta1.RouteConditionResolvedRefs, metav1.ConditionFalse, status.ReasonDegraded, "At least one Spec.Rules.BackendRef must be specified.")
+		return nil, totalWeight, false
+	}
+
+	var clusters []*Cluster
+
+	// Validate the backend refs.
+	for _, backendRef := range backendRefs {
+		service, cond := p.validateBackendRef(backendRef.BackendRef, KindGRPCRoute, routeNamespace)
+		if cond != nil {
+			routeAccessor.AddCondition(gatewayapi_v1beta1.RouteConditionType(cond.Type), cond.Status, gatewayapi_v1beta1.RouteConditionReason(cond.Reason), cond.Message)
+			continue
+		}
+
+		var clusterRequestHeaderPolicy, clusterResponseHeaderPolicy *HeadersPolicy
+		for _, filter := range backendRef.Filters {
+			switch filter.Type {
+			case gatewayapi_v1alpha2.GRPCRouteFilterRequestHeaderModifier:
+				// Per Gateway API docs, "specifying a core filter multiple times has
+				// unspecified or custom conformance.", here we choose to just process
+				// the first one.
+				if clusterRequestHeaderPolicy != nil {
+					continue
+				}
+
+				var err error
+				clusterRequestHeaderPolicy, err = headersPolicyGatewayAPI(filter.RequestHeaderModifier, string(filter.Type))
+				if err != nil {
+					routeAccessor.AddCondition(gatewayapi_v1beta1.RouteConditionResolvedRefs, metav1.ConditionFalse, status.ReasonDegraded, fmt.Sprintf("%s on request headers", err))
+				}
+			case gatewayapi_v1alpha2.GRPCRouteFilterResponseHeaderModifier:
+				// Per Gateway API docs, "specifying a core filter multiple times has
+				// unspecified or custom conformance.", here we choose to just process
+				// the first one.
+				if clusterResponseHeaderPolicy != nil {
+					continue
+				}
+
+				var err error
+				clusterResponseHeaderPolicy, err = headersPolicyGatewayAPI(filter.ResponseHeaderModifier, string(filter.Type))
+				if err != nil {
+					routeAccessor.AddCondition(gatewayapi_v1beta1.RouteConditionResolvedRefs, metav1.ConditionFalse, status.ReasonDegraded, fmt.Sprintf("%s on response headers", err))
+				}
+			default:
+				routeAccessor.AddCondition(status.ConditionNotImplemented, metav1.ConditionTrue, status.ReasonHTTPRouteFilterType, "GRPCRoute.Spec.Rules.BackendRef.Filters: Only RequestHeaderModifier and ResponseHeaderModifier type is supported.")
+			}
+		}
+
+		// Route defaults to a weight of "1" unless otherwise specified.
+		routeWeight := uint32(1)
+		if backendRef.Weight != nil {
+			routeWeight = uint32(*backendRef.Weight)
+		}
+
+		// Keep track of all the weights for this set of backend refs. This will be
+		// used later to understand if all the weights are set to zero.
+		totalWeight += routeWeight
+
+		// https://github.com/projectcontour/contour/issues/3593
+		service.Weighted.Weight = routeWeight
+		clusters = append(clusters, &Cluster{
+			Upstream:              service,
+			Weight:                routeWeight,
+			Protocol:              service.Protocol,
```

Comment:
We probably want to make a little change here:
- The protocol here comes from the referenced backend Service annotations if they are set (see: https://github.com/projectcontour/contour/blob/19a4c9f0a372838782d30ffa78c3c261cfaf6f34/internal/dag/accessors.go#L64)
- If the Service annotations are set, we should probably respect them, for backwards compatibility etc.
- If not set, currently we will configure a HTTP/1 cluster 
  - https://github.com/projectcontour/contour/blob/19a4c9f0a372838782d30ffa78c3c261cfaf6f34/internal/envoy/v3/cluster.go#L115
  - https://github.com/projectcontour/contour/blob/19a4c9f0a372838782d30ffa78c3c261cfaf6f34/internal/envoy/v3/cluster.go#L308

I think if the protocol is not set on the Service, we should maybe assume a protocol that matches what listener the route was attached to, e.g. `h2` for HTTPS listener and `h2c` for HTTP, given the user is explicitly asking for a gRPC route not a general HTTP route